UnityProject/Assets/Scripts/Items/Implants/Organs/Liver.cs

```diff
@@ -66,70 +66,98 @@ public override void ImplantPeriodicUpdate()
 			float drawnAmount = 0;
 
 			//figure out how much we are going to process or remove
-			foreach (Reagent reagent in blood.CurrentReagentMix.reagents.Keys)
+			lock (blood.CurrentReagentMix.reagents)
 			{
-				if (Alcohols.AlcoholicReagents.Contains(reagent) || Toxins.Contains(reagent))
+				foreach (Reagent reagent in blood.CurrentReagentMix.reagents.Keys)
 				{
-					float amount = Mathf.Min(tickPullProcessingAmnt,RelatedPart.BloodContainer.CurrentReagentMix[reagent]);
-					amount = Mathf.Min(amount, (processingContainer.MaxCapacity - processingContainer.ReagentMixTotal)-drawnAmount);
+					if (Alcohols.AlcoholicReagents.Contains(reagent) || Toxins.Contains(reagent))
+					{
+						float amount = Mathf.Min(tickPullProcessingAmnt,RelatedPart.BloodContainer.CurrentReagentMix[reagent]);
+						amount = Mathf.Min(amount, (processingContainer.MaxCapacity - processingContainer.ReagentMixTotal)-drawnAmount);
 
-					tempArray.Add(new Tuple<Reagent, float>(reagent, amount));
+						tempArray.Add(new Tuple<Reagent, float>(reagent, amount));
 
-					if (processingContainer.IsFull)
-					{
-						Logger.LogTrace("Liver is full, please try again. or don't.",Category.Health);
-						break;
-					}
+						if (processingContainer.IsFull)
+						{
+							Logger.LogTrace("Liver is full, please try again. or don't.",Category.Health);
+							break;
+						}
 
-					drawnAmount += amount;
-					tickPullProcessingAmnt -= amount;
-					if (tickPullProcessingAmnt <= 0) break;
+						drawnAmount += amount;
+						tickPullProcessingAmnt -= amount;
+						if (tickPullProcessingAmnt <= 0) break;
+					}
 				}
+
 			}
 
+			List<Tuple<Reagent,float>> Toremove = new List<Tuple<Reagent,float>>();
```

Comment:
this and the extra foreachs you presented aren't needed, that's adding a whole other temporary data structure on top of a temporary data structure. If temp array is having issues being removed from in the foreach, then i'll just change it so the temparray is being cleared afterwards. A bit miffed i wasn't awake to see this pr before it was merged.

Reply:
wouldn't iterating backwards be good enough to solve the issue?

Reply:
it's not an issue of how it's iterating, just an issue of what temparray is used for. it's simple enough to just clear temparray after the foreachs that read through it, which i'm gonna pull up a pr to do just that.